### ISSUE TITLE
Torchrun does not handle worker failure gracefully

### ISSUE THREAD

**Issue by shravan-achar** [pytorch/pytorch#146371]
### 🐛 Describe the bug

```
def train_func():
    import os
    import torch.distributed as dist
    import time
    import sys

    dist.init_process_group(backend="nccl")

    ws = dist.get_world_size()
    rank = dist.get_rank()
    endpoint = os.getenv("PET_RDZV_ENDPOINT")

    print(f"WS: {ws}, RANK: {rank}")
    print(endpoint)

    dist.barrier()

    for i in range(40):
        print(i)
        time.sleep(1)
        if i > 10 and rank == 1:
            sys.exit(1)
```

This function is converted into a script. This is the command used,

```
torchrun <converted_script.py>

Env variables:

      PET_RDZV_ENDPOINT= test-torch-restart-master-0:23456
      PYTHONUNBUFFERED= 1
      MASTER_PORT=23456
      PET_MASTER_PORT=23456
      MASTER_ADDR=test-torch-restart-master-0
      PET_MASTER_ADDR=test-torch-restart-master-0
      WORLD_SIZE=3
      RANK=0
      PET_NODE_RANK=0
      PET_NPROC_PER_NODE=1
      PET_NNODES=3
```

We expected that when worker process with rank 1 failed, all of the workers must be restarted by torchrun. We are not seeing that. What we see is that worker just fails while the other workers continue to run (including the master). Is that accurate with the expectations of how torchrun is supposed to work with handling of worker failures? Also, the failed worker does not get restarted.  

We are running on Vanilla Kubernetes (Linux and x86_64)

### Versions

torchrun version 2.1.2

**Comment by shravan-achar** [comment]
cc: @andreyvelich

**Comment by Unturned3** [comment]
I also see this behavior in a multi-node `torchrun` setup on SLURM. If I `kill -9` a child process of `torchrun` on one node, instances on other nodes continue running until SLURM kills them. I was expecting `torchrun` to communicate amongst each other & either all fail or restart everyone.

**Comment by fduwjj** [comment]
@Unturned3 yeah it is indeed expected to have scheduler to handle the termination and communication like this, meaning in the case of multi-node torchrun does not kill it for you.

**Comment by vHitsuji** [comment]
@Unturned3 
I might be wrong but I think in this case the local Torchrun to the rank that fails does not even kill its own children.
At least this is what I observe.
